Mixpanel/ReadWriteLock.swift

```diff
@@ -0,0 +1,31 @@
+//
+//  ReadWriteLock.swift
+//  Mixpanel
+//
+//  Created by Hairuo Sang on 8/9/17.
+//  Copyright © 2017 Mixpanel. All rights reserved.
+//
+
+import Foundation
+
+class ReadWriteLock{
+    //A placeholder object that tracking/archiving/flushing threads lock on
+    let conccurentQueue: DispatchQueue
+    
+    init(label: String) {
+        self.conccurentQueue = DispatchQueue(label: label, attributes: .concurrent)
+    }
+    
+    func read(closure: () -> ()) {
+        self.conccurentQueue.sync {
+            closure();
+        }
+    }
+    
+    func write(closure: () -> ()) {
+        self.conccurentQueue.sync(flags: .barrier, execute: {
```

Comment:
this is probably fine. i don't know why apple didn't document `.barrier` https://developer.apple.com/documentation/dispatch/dispatchworkitemflags/1780674-barrier?changes=latest_minor but i assume it works the same as https://developer.apple.com/documentation/dispatch/1452917-dispatch_barrier_sync